### PR TITLE
Introduce cli

### PR DESCRIPTION
**Pull request by apostergiou**
Usage:

```shell
$ go run client/main.go build --host localhost:9090 --project assets --group group_name \
--target /tmp/synced_assets -- --gemfile=@/home/apostergiou/Gemfile --foo=bar

# Check the HTTP request
$ nc -l -p 9090
```

### CODE REVIEW

File: client/main.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      directory /tmp/assets and uses the dynamic arguments gemfile and foo.
+      Note the @ before the Gemfile, this indicated a referral to an actual file on the
+      filesystem.
+
+      $ mistryctl build --host example.org:9090 --project rails-assets \
+      --group group_name --target /tmp/assets -- --gemfile=@Gemfile --foo=bar
+
+   2. The following secuence uses the GROUP environment variable set the project group.
+
+      $ GROUP=skroutz mistryctl build --host example.org:9090 --project rails-assets \
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: True, added.

File: client/main.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+   1. The following sequence schedules a command to be executed in host example.org:9090,
+      builds the rails-assets project, groups it under group_name, syncs the result to the
+      directory /tmp/assets and uses the dynamic arguments gemfile and foo.
+      Note the @ before the Gemfile, this indicated a referral to an actual file on the
+      filesystem.
+
+      $ mistryctl build --host example.org:9090 --project rails-assets \
+      --group group_name --target /tmp/assets -- --gemfile=@Gemfile --foo=bar
+
+   2. The following secuence uses the GROUP environment variable set the project group.
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: :+1:

File: client/main.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	cli.AppHelpTemplate = fmt.Sprintf(`%s
+WEBSITE: https://github.com/skroutz/mistry
+`, cli.AppHelpTemplate)
+
+	cli.CommandHelpTemplate = fmt.Sprintf(`%s
+DYNAMIC OPTIONS:
+   -- dynamic options for the command
+
+EXAMPLES:
+   1. The following sequence schedules a command to be executed in host example.org:9090,
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: I think we could use `job` here. Updated.

File: client/main.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      builds the rails-assets project, groups it under group_name, syncs the result to the
+      directory /tmp/assets and uses the dynamic arguments gemfile and foo.
+      Note the @ before the Gemfile, this indicated a referral to an actual file on the
+      filesystem.
+
+      $ mistryctl build --host example.org:9090 --project rails-assets \
+      --group group_name --target /tmp/assets -- --gemfile=@Gemfile --foo=bar
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: We could use `yarn`.

File: client/main.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	app.Usage = "Runs commands inside pre-defined environments and makes their output for consumption"
+	app.HideVersion = true
+	app.Commands = []cli.Command{
+		{
+			Name:  "build",
+			Usage: "Runs a command and syncs its output",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:        "host",
+					Usage:       "execution environment host",
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: fixed

File: client/main.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+					Destination: &host,
+					EnvVar:      "HOST",
+				},
+				cli.StringFlag{
+					Name:        "project, p",
+					Usage:       "namespace for the execution environment",
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: updated

File: client/main.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+					Usage:       "the local directory where the result will be saved",
+					Destination: &target,
+					EnvVar:      "TARGET",
+				},
+			},
+			Action: func(c *cli.Context) {
+
+				// Validate existence of mandatory arguments
+				if host == "" {
+					log.Fatal("host cannot be empty")
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: True, updated to return errors as appropriate.

File: client/main.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+					}
+				}
+				paramsMap, err := json.Marshal(params)
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				// Communicate the command to the HTTP server.
+				reader := strings.NewReader(fmt.Sprintf(`{"project": "%s", "group": "%s", "params": %s}`, project, group, paramsMap))
+				request, err := http.NewRequest("POST", fmt.Sprintf("http://%s/%s", host, jobsPath), reader)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: fixed

File: client/main.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+					if err != nil {
+						log.Fatal(err)
+					}
+
+					err = json.Unmarshal([]byte(body), &buildResult)
+					if err != nil {
+						log.Fatal(err)
+					}
+
+					cmd := exec.Command("rsync", "-rlpt", buildResult.Path, target)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Seems reasonable enough. I adjusted the code accordingly.

File: client/main.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+				}
+
+				// Sync the result to the target for 201 responses.
+				if resp.StatusCode == http.StatusCreated {
+					body, err := ioutil.ReadAll(resp.Body)
+					if err != nil {
+						log.Fatal(err)
+					}
+
+					err = json.Unmarshal([]byte(body), &buildResult)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Added a check to handle the response type, which, as you mentioned, is going to be `rsync` for now.

File: Gopkg.toml [hunk 11]
```diff
@@ -29,3 +29,7 @@
   version = "17.05.0-ce"
 	name = "github.com/docker/docker"
 
+[[constraint]]
+  version = "1.19.1"
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply Gopkg.toml]: True. Either way, we have already added the dependency in https://github.com/skroutz/mistry/commit/6fc978e89701d83b3cd7b7a026baff52cb1970ae, so I dropped the commit from this pull request.

File: client/main.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+					log.Fatal(err)
+				}
+				client := &http.Client{}
+				resp, err := client.Do(request)
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				// Sync the result to the target for 201 responses.
+				if resp.StatusCode == http.StatusCreated {
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Good catch, updated.

File: client/main.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+			},
+		},
+	}
+
+	err := app.Run(os.Args)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	os.Exit(0)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Interesting discovery here, thanks!

File: client/main.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+						params[k] = string(data)
+					}
+				}
+				paramsMap, err := json.Marshal(params)
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				// Communicate the command to the HTTP server.
+				reader := strings.NewReader(fmt.Sprintf(`{"project": "%s", "group": "%s", "params": %s}`, project, group, paramsMap))
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Fixed.

File: client/main.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+				request, err := http.NewRequest("POST", fmt.Sprintf("http://%s/%s", host, jobsPath), reader)
+				if err != nil {
+					log.Fatal(err)
+				}
+				client := &http.Client{}
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: A legitimate concern. Nevertheless, let's tackle this in a separate pull request.

File: worker_test.go [hunk 16]
```diff
@@ -122,54 +146,100 @@ func TestExitCode(t *testing.T) {
...
 func TestResultCache(t *testing.T) {
-	result1, err := postJob(JobRequest{"result-cache", params, ""})
+	target, err := ioutil.TempDir("", "mistry-tests-results")
+	if err != nil {
+		log.Fatal(err)
+	}
+	user, err := user.Current()
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (agis) [inline worker_test.go]: This is needed in various tests, so you could do this once in `TestMain()`

File: worker_test.go [hunk 17]
```diff
@@ -337,3 +408,13 @@ func waitForServer(port string) {
 	}
 	log.Fatalf("Server on port %s not up after 10 retries", port)
 }
+
+func build(args ...string) (string, error) {
```
Comment (agis) [inline worker_test.go]: This function only really needs the following arguments:

- project
- group
- job parameters

The others could be extracted to constants.

File: .gitignore [hunk 18]
```diff
@@ -6,5 +6,6 @@
 config.json
 fabfile.py
 mistry
+client/client
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply .gitignore]: Fixed.

File: client/main.go [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+					Value:       ".",
+				},
+				cli.StringFlag{
+					Name:        "transport",
+					Destination: &transport,
+					Value:       "scp",
+				},
+			},
+			Action: func(c *cli.Context) error {
+
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: Removed.

File: client/main.go [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+					if strings.HasPrefix(v, "@") {
+						data, err := ioutil.ReadFile(strings.TrimPrefix(v, "@"))
+						if err != nil {
+							return err
+						}
+						params[k] = string(data)
+					}
+				}
+
+				// Communicate the command to the HTTP server.
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: ok, updated.

File: client/main.go [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+				resp, err := client.Do(request)
+				if err != nil {
+					return err
+				}
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					return err
+				}
+
+				// Any non 201 responses are considered failed.
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply client/main.go]: done

File: worker_test.go [hunk 22]
```diff
@@ -56,64 +60,84 @@ func TestMain(m *testing.M) {
 }
 
 func TestSimpleBuild(t *testing.T) {
-	jr := JobRequest{Project: "simple", Params: params, Group: ""}
-	res, err := postJob(jr)
+	target, err := ioutil.TempDir("", "mistry-tests-results")
 	if err != nil {
-		t.Fatal(err)
+		panic(err)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply worker_test.go]: This slipped through my review. I added a panic here as a shortcut during the test development.

File: worker_test.go [hunk 23]
```diff
@@ -122,54 +146,100 @@ func TestExitCode(t *testing.T) {
...
+	if err != nil {
+		log.Fatalf("Error output: %s, err: %v", cmdOut1, err)
+	}
+	out1, err := ioutil.ReadFile(filepath.Join(target, ArtifactsDir, "out.txt"))
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	out2, err := readOut(result2, ArtifactsDir)
+	cmdOut2, err := build("build", "--host", "localhost", "--port", "8462", "--project", "result-cache", "--target", target, "--transport-user", user.Username, "--group", "foo", "--", "--foo=bar2")
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply worker_test.go]: True, extracted to a constant.

File: worker_test.go (hunk 16 above)
Comment (apostergiou) [reply worker_test.go]: True, extracted in `TestMain()` and `init()` as appropriate.

File: worker_test.go (hunk 17 above)
Comment (apostergiou) [reply worker_test.go]: Indeed, we can extract `host`, `port` to constants and also utilize the global variables `target` and `username`.

So, in the end this becomes `func build(project, group, params string) (string, error) {`

File: worker_test.go [hunk 24]
```diff
@@ -45,75 +60,79 @@ func TestMain(m *testing.M) {
...
-
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		var err error
-		result1, err = postJob(jr)
+		cmdOut, err := build("--project", "build-coalescing", "--group", "foo")
 		if err != nil {
-			log.Fatal(err)
+			t.Fatalf("Error output: %s, err: %v", cmdOut, err)
```
Comment (agis) [inline worker_test.go]: Only the main goroutine should call `t.Fatal` per https://golang.org/pkg/testing/#T.FailNow.

This should remain as it was.

Comment (apostergiou) [reply worker_test.go]: Fixed, thanks for clarifying this.

**Pull request merged**